### ISSUE TITLE
MobileFacenet SE Train from scratch

### ISSUE THREAD

**Issue by malik-anhar** [leondgarse/Keras_insightface#9]
I just want to know how do you train the se_mobile_facenet. May I know the parameter you use to train the mobile facenet model? (Optimizer, learning rate, batch size, loss function and its alpha, dataset, number of epoch for each scheduler or anything else). Thank you in advance

**Comment by leondgarse** [comment]
It's too far away, and most of those things seems wrong now...
As far as I can remeber and from my recorded data, it's configured as this:
```py
import losses, train

basic_model = mobile_facenet.mobile_facenet(256, dropout=0, name="se_mobile_facenet_256", use_se=True)
data_path = '/datasets/faces_emore_112x112_folders'
eval_paths = ['/datasets/faces_emore/lfw.bin', '/datasets/faces_emore/cfp_fp.bin', '/datasets/faces_emore/agedb_30.bin']

tt = train.Train(data_path, save_path='keras_se_mobilefacenet_emore.h5', eval_paths=eval_paths,
    basic_model=basic_model, lr_base=0.001, batch_size=640, random_status=3)
sch = [
  {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "epoch": 15, "optimizer": "nadam"},
  {"loss": losses.ArcfaceLoss(scale=64), "epoch": 4, "bottleneckOnly": True},
  {"loss": losses.ArcfaceLoss(scale=64), "epoch": 30},
  {"loss": losses.BatchHardTripletLoss(0.35), "epoch": 10},
  {"loss": losses.BatchHardTripletLoss(0.3), "epoch": 10},
  {"loss": losses.BatchHardTripletLoss(0.25), "epoch": 10},
  {"loss": losses.BatchHardTripletLoss(0.2), "epoch": 20},
]
tt.train(sch, 0)
```

| Loss               | Epochs | First epoch (label_smoothing=0.1)                   |
| ------------------ | ------ | --------------------------------------------------- |
| Softmax            | 15     | 13256s 2s/step - loss: 5.9982 - accuracy: 0.3615    |
| Bottleneck Arcface | 4      | 4111s 452ms/step - loss: 21.7145 - accuracy: 0.8624 |
| Arcface 64         | 30     | 13043s 1s/step - loss: 16.7003 - accuracy: 0.9491   |
| Triplet (BS 1440)  | 50     | 6688s 2s/step - loss: 0.2319  |

- When looking back at this
  - It's better to use an optimizer with `weight_decay`, like `SGDW` / `AdamW`.
  - `bottleneckOnly` is NOT essential here, as I will reload the weights data from `softmax` to `arcface`.
  - Maybe train `Arcface` from scratch without `softmax` is better.
  - Inferencing `se_mobile_facenet` is too slow, though it's smaller.
- Recently I'm mainly using `SGDW`, I would recommend this simple one but NOT tested:
  ```py
  import tensorflow_addons as tfa
  import train, losses

  basic_model = mobile_facenet.mobile_facenet(256, dropout=0, name="se_mobile_facenet_256", use_se=True)
  data_path = '/datasets/faces_emore_112x112_folders'
  eval_paths = ['/datasets/faces_emore/lfw.bin', '/datasets/faces_emore/cfp_fp.bin', '/datasets/faces_emore/agedb_30.bin']
  tt = train.Train(data_path, save_path='keras_se_mobilefacenet_emore.h5',
      eval_paths=eval_paths, basic_model=basic_model, model=None, lr_base=0.1, lr_decay=0.1, lr_decay_steps=[20, 30],
      batch_size=512, random_status=0, output_wd_multiply=10)

  optimizer = tfa.optimizers.SGDW(learning_rate=0.1, weight_decay=5e-4, momentum=0.9)
  sch = [
      {"loss": losses.ArcfaceLossT4(scale=64), "epoch": 40},
  ]
  tt.train(sch, 0)
  ```
- And the dataset, Insightface have uploaded a new dataset [glint360k](https://github.com/deepinsight/insightface/tree/master/recognition/partial_fc#glint360k), which seems can reach a better result, but it's too large...

**Comment by malik-anhar** [comment]
Thanks for the answer, I'll try it as soon as possible. But I found an issue that a model trained with Triplet loss has smaller cosine distance (when I do inference for the same person) and larger cosine distance (for different person) compared to model that trained with ArcFace loss. Is there any intuition why this is happening?

**Comment by leondgarse** [comment]
Ya, I found that too. There actually are some evidences in the paper [ArcFace: Additive Angular Margin Loss for Deep Face Recognition](https://arxiv.org/pdf/1801.07698.pdf) section `3.2. Ablation Study on Losses`.

![Selection_281](https://user-images.githubusercontent.com/5744524/96821608-d8181080-145a-11eb-91ad-8906b1b7d55b.png)

I tried to combine `triplet` with `arcface`, and lower `triplet alpha` every `10` epochs, it seems better.
```py
sch = [
    {"loss": losses.ArcfaceLoss(scale=64), "epoch": 10, "triplet": 32, "alpha": 0.35},
    {"loss": losses.ArcfaceLoss(scale=64), "epoch": 10, "triplet": 32, "alpha": 0.3},
    {"loss": losses.ArcfaceLoss(scale=64), "epoch": 10, "triplet": 32, "alpha": 0.25},
    {"loss": losses.ArcfaceLoss(scale=64), "epoch": 10, "triplet": 32, "alpha": 0.2},
]
```

**Comment by malik-anhar** [comment]
I see, so it's better to train the model by combining the ```arcface``` with ```triplet``` loss to get the smaller cosine distance. Or I can train the model with ```arcface``` first, then continue the training by using ```BatchHardTripletLoss```. And if the model won't convergence, I can reduce the margin, right?

**Comment by leondgarse** [comment]
Yes. The author of the `ArcFace paper` also tried that combination, but found no good, so I think it's better to try `arcface` only first.

**Comment by malik-anhar** [comment]
I'll try it, thank you
